akka-stream-tests/src/test/scala/akka/stream/scaladsl/RetryFlowSpec.scala

```diff
@@ -0,0 +1,572 @@
+
+      // let element go via retryable flow
+      val try1 = internalOut.requestNext()
+      try1._1 shouldBe element1._1
+      try1._2 shouldBe element1._2
+      internalIn.sendNext(Failure(new RuntimeException("boom")) -> 1)
+
+      // let element go via retryable flow
+      val try2 = internalOut.requestNext(3.seconds)
+      try2._1 shouldBe "A'"
+      try2._2 shouldBe 2
+      internalIn.sendNext(Success("Ares") -> 1)
+
+      // expect result
+      externalOut.requestNext() shouldBe Success("Ares") -> 1
+    }
+
+    "allow to send two elements" in new ConstructBench[InData, Ctx2, OutData]((_, out) =>
+      out._1 match {
+        case Failure(_) => Some("A'" -> 2)
+        case Success(_) => None
+      }) {
+      // demand on stream end
+      externalOut.request(2)
+
+      // push element
+      val element1 = "A" -> 1
+      externalIn.sendNext(element1)
+      val element2 = "B" -> 2
+      externalIn.sendNext(element2)
+
+      // let element go via retryable flow
+      val try1 = internalOut.requestNext()
+      try1._1 shouldBe element1._1
+      try1._2 shouldBe element1._2
+      internalIn.sendNext(Failure(new RuntimeException("boom")) -> 1)
+
+      // let element go via retryable flow
+      val try2 = internalOut.requestNext()
+      try2._1 shouldBe "A'"
+      try2._2 shouldBe 2
+      internalIn.sendNext(Success("Ares") -> 1)
+
+      // let element2 go via retryable flow
+      val try2_1 = internalOut.requestNext()
+      try2_1._1 shouldBe element2._1
+      try2_1._2 shouldBe element2._2
+      internalIn.sendNext(Success("Bres") -> 1)
+
+      // expect result
+      externalOut.requestNext() shouldBe Success("Ares") -> 1
+      externalOut.requestNext() shouldBe Success("Bres") -> 1
+    }
+
+    "fail if inner flow emits twice" in new ConstructBench[InData, Ctx2, OutData]((_, _) => None) {
+      externalOut.request(99)
+
+      // push element
+      val elA = "A" -> 123
+      externalIn.sendNext(elA)
+
+      // let element go via retryable flow
+      val elA2 = internalOut.requestNext()
+      elA2._1 shouldBe elA._1
+      elA2._2 shouldBe elA._2
+      internalIn.sendNext(Success("result A") -> 123)
+      internalIn.sendNext(Success("result B") -> 222)
+
+      // expect result
+      externalOut.requestNext() shouldBe Success("result A") -> 123
+      externalOut.expectError() shouldBe an[IllegalStateException]
+    }
+
+    // TODO
+    // Not sure if it should protect against inner flows not emitting elements
+    "fail if inner flow filters elements" ignore new ConstructBench[InData, Ctx2, OutData]((_, _) => None) {
+      externalOut.request(99)
+
+      // push element
+      val elA = "A" -> 123
+      externalIn.sendNext(elA)
+
+      // let element go via retryable flow
+      val elA2 = internalOut.requestNext()
+      elA2._1 shouldBe elA._1
+      elA2._2 shouldBe elA._2
+
+      // expect result
+      externalOut.expectError() shouldBe a[TimeoutException]
+    }
+
+  }
+
+}
```

Comment:
Since this operator is detached in some meaning, I think there is a missing test case that verifies that it does not pull from upstream until the element has reached downstream.